File: app/src/main/java/com/sunbuy/qrcardid_java/scan/QRCodeImageGallery.java
package com.sunbuy.qrcardid_java.scan;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.mlkit.vision.barcode.BarcodeScanner;
import com.google.mlkit.vision.barcode.BarcodeScannerOptions;
import com.google.mlkit.vision.barcode.BarcodeScanning;
import com.google.mlkit.vision.barcode.common.Barcode;
import com.google.mlkit.vision.common.InputImage;

import java.util.List;

public class QRCodeImageGallery {
    private int[] barcodeFormats;
    private QRCodeAnalyzerCallback callback;
    private BarcodeScanner barcodeScanner;
    private BarcodeScannerOptions.Builder optionsBuilder;

    public QRCodeImageGallery(int[] barcodeFormats) {
        this.barcodeFormats = barcodeFormats;
        if (barcodeFormats.length > 1) {
            optionsBuilder = new BarcodeScannerOptions.Builder().setBarcodeFormats(Barcode.FORMAT_QR_CODE, Barcode.FORMAT_UPC_A);
        } else {
            optionsBuilder = new BarcodeScannerOptions.Builder().setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS);
        }
    }

    public void setQRCodeAnalyzerCallback(QRCodeAnalyzerCallback callback) {
        this.callback = callback;

        try {
            barcodeScanner = BarcodeScanning.getClient(optionsBuilder.build());
        } catch (Exception exception) {
            callback.onFailure(exception);
        }
    }

    public void onProcess(InputImage inputImage) {
        if (barcodeScanner != null) {
            barcodeScanner.process(inputImage)
                    .addOnSuccessListener(barcodes -> {
                        if (barcodes.size() > 0) {
                            if (barcodes.get(0) != null) {
                                callback.onSuccess(barcodes.get(0));
                            } else {
                                callback.notFound();
                            }
                        }else {
                            callback.notFound();
                        }
                    }).addOnFailureListener(e -> {
                        callback.onFailure(e);
                    });
        }
    }
}

File: scancode/src/main/java/com/sunbuy/scancode/QRCodeImageGallery.kt
package com.sunbuy.scancode

import android.content.Context
import android.net.Uri
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage

class QRCodeImageGallery(
    private val barcodeFormats: IntArray,
    private val onSuccess: ((Barcode) -> Unit),
    private val notFound: (() -> Unit),
    private val onFailure: ((Exception) -> Unit),
) {

    private val barcodeScanner by lazy {
        val optionsBuilder = if (barcodeFormats.size > 1) {
            BarcodeScannerOptions.Builder()
                .setBarcodeFormats(barcodeFormats.first(), *barcodeFormats.drop(1).toIntArray())
        } else {
            BarcodeScannerOptions.Builder()
                .setBarcodeFormats(barcodeFormats.firstOrNull() ?: Barcode.FORMAT_UNKNOWN)
        }
        try {
            BarcodeScanning.getClient(optionsBuilder.build())
        } catch (e: Exception) { // catch if for some reason MlKitContext has not been initialized
            onFailure(e)
            null
        }
    }

    fun onProcess(inputImage: InputImage) {
        barcodeScanner?.let { scanner ->
            scanner.process(inputImage)
                .addOnSuccessListener { codes ->
                    if (!codes.isNullOrEmpty()) {
                        codes.firstNotNullOfOrNull { it }?.let { onSuccess(it) }
                    } else {
                        notFound.invoke()
                    }
                }
                .addOnFailureListener {
                    onFailure(it)
                }
        }
    }
}